userbenchmark/ddp_experiments/__init__.py

```diff
@@ -134,26 +143,95 @@ def get_init_file(args):
+        self.store.add(my_key, 1)
+        wait_for = []
+        for i in range(self.world_size):
+            key = f"barrier{self.call_idx}.{i}"
+            wait_for.append(key)
+        print(f"waiting on {wait_for}")
+        self.store.wait(wait_for)
+
+
+@dataclasses.dataclass
+class JobConfig:
+    outer_sync_path: str
 
-    def __call__(self):
-        self._setup_gpu_args()
 
-        pos = self.args.model.rfind(".")
-        module = importlib.import_module(self.args.model[:pos])
-        model_class = getattr(module, self.args.model[(pos+1):])
-
-        pos = self.args.trainer.rfind(".")
-        module = importlib.import_module(self.args.trainer[:pos])
-        trainer_class = getattr(module, self.args.trainer[(pos+1):])
+class TrainerWrapper(object):
+    # per_experiment_args is a list of expriments.
+    # Each experiment should be a tuple of (config dict, args, model_args).
+    # config: configuration data to attach to the result dict.
+    # args & model_args: arguments for core_model.Trainer.
+    def __init__(self, job_config: JobConfig, per_experiment_args: List[Tuple[Dict, Any, Any]]):
+        self.job_config = job_config
+        self.per_experiment_args = per_experiment_args
+
+    # this is called within a multiprocessing.Process.
+    def run_once(self, args, model_args, q):
+        print("run_once")
+        self._setup_gpu_args(args)
+
+        pos = args.model.rfind(".")
+        module = importlib.import_module(args.model[:pos])
+        model_class = getattr(module, args.model[(pos+1):])
+
+        pos = args.trainer.rfind(".")
+        module = importlib.import_module(args.trainer[:pos])
+        trainer_class = getattr(module, args.trainer[(pos+1):])
+
+        trainer = trainer_class(args, model_class, model_args=model_args)
+        result = trainer.measure()
+        print(f"result {result}")
+        q.put(result)
+        trainer.teardown()
 
-        return trainer_class(self.args, model_class, model_args=self.model_args).measure()
+    def __call__(self):
+        results = []
+        job_env = submitit.JobEnvironment()
+        barrier = self._get_barrier()
+        print(f"This is node {job_env.node}")
+        for (config, args, model_args) in self.per_experiment_args:
+            try:
+                if job_env.node >= args.nodes:
+                    continue
+                result_dict = {**config}
+                q = multiprocessing.Queue()
+                proc = multiprocessing.Process(target=self.run_once, args=(args, model_args, q))
+                proc.start()
+                proc.join()
+                print(f"exit code: {proc.exitcode} and result: {result_dict}")
+                if proc.exitcode == 0:
+                    result_dict['result'] = q.get()
+                else:
+                    result_dict['result'] = None
+                assert 'result' in result_dict
+                # wrap in <RESULT></RESULT> so we can parse partial results in the stdout logs
```

Comment:
this is fine, but is it not possible to just monitor some path on /fsx and see partial results show up as individual files there as they are written?

Reply:
we can already do that, but I forgot to include the changes to parse_ddp.py in this PR - I'll open another PR with those changes.